File: leetcode/454.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
 public:
  int fourSumCount(vector<int>& A, vector<int>& B, vector<int>& C, vector<int>& D) {
    unordered_map<int, int> cdMap;
    for (auto c = C.begin(); c != C.end(); c++) {
      for (auto d = D.begin(); d != D.end(); d++) {
        cdMap[*c + *d]++;
      }
    }
    int ret = 0;
    for (auto a = A.begin(); a != A.end(); a++) {
      for (auto b = B.begin(); b != B.end(); b++) {
        ret += cdMap[-*a - *b];
      }
    }
    return ret;
  }
};

File: leetcode/454.js
var fourSumCount = function(A, B, C, D) {
  const cdMap = new Map();
  C.forEach(c => {
    D.forEach(d => {
      const sum = c + d;
      const count = cdMap.get(sum);
      cdMap.set(sum, 1 + (count ? count : 0));
    });
  });

  let ret = 0;
  A.forEach(a => {
    B.forEach(b => {
      const target = -(a + b);
      const count = cdMap.get(target);
      ret += count ? count : 0;
    });
  });
  return ret;
};
